diff --git a/fs/devpts/inode.c b/fs/devpts/inode.c
index f1394c691..d3c791f1e 100644
--- a/fs/devpts/inode.c
+++ b/fs/devpts/inode.c
@@ -612,16 +612,6 @@ extern int ksu_handle_devpts(struct inode*);
  */
 void *devpts_get_priv(struct dentry *dentry)
 {
-#ifdef CONFIG_KSU_SUSFS_SUS_SU
-	if (likely(susfs_is_current_proc_su_not_allowed())) {
-		goto orig_flow;
-	}
-	if (likely(ksu_devpts_hook)) {
-		ksu_handle_devpts(dentry->d_inode);
-	}
-orig_flow:
-#endif
-
 	if (dentry->d_sb->s_magic != DEVPTS_SUPER_MAGIC)
 		return NULL;
 	return dentry->d_fsdata;
