diff --git a/kernel/selinux/rules.c b/kernel/selinux/rules.c
index 55613af5..4f3b07fe 100644
--- a/kernel/selinux/rules.c
+++ b/kernel/selinux/rules.c
@@ -26,11 +26,11 @@ static struct policydb *get_policydb(void)
 
 static DEFINE_MUTEX(ksu_rules);
 
-void apply_kernelsu_rules()
+void ksu_apply_kernelsu_rules()
 {
 	struct policydb *db;
 
-	if (!getenforce()) {
+	if (!ksu_getenforce()) {
 		pr_info("SELinux permissive or disabled, apply rules!\n");
 	}
 
