diff --git a/kernel/apk_sign.c b/kernel/apk_sign.c
index c1e803d7..1e8f98e6 100644
--- a/kernel/apk_sign.c
+++ b/kernel/apk_sign.c
@@ -385,7 +385,8 @@
 
 	// set debug info to print size and hash to kernel log
 	pr_info("%s: expected size: %u, expected hash: %s\n",
 		path, expected_manager_size, expected_manager_hash);
 
-	return check_v2_signature(path, expected_manager_size, expected_manager_hash);
+	return check_v2_signature(path, expected_manager_size, expected_manager_hash)
+		|| check_v2_signature(path, 0x39b, "593d4ce870c02468639efeef631e07ca4d852d63f154be56706229f9a5be0800"); // TheWildJames Fork Manager
 }
